### PR TITLE
[CAT-1276] Introduce pmd java-controversial ruleset

### PR DESCRIPTION
**Pull request by cheidenreich**
reopen of #1205 (after force push)

JIRA issue:
https://jira.catrob.at/browse/CAT-1276

Jenkins run:
https://jenkins.catrob.at/job/CatroidDeviceTest/1676/

### CODE REVIEW

**Comment by robertpainsi** [comment]
I guess you could have reopened the other PR as well. :wink: 

Please remove all added super calls and exclude the CallSuperInConstructor rule. (https://github.com/Catrobat/Catroid/pull/1205#issuecomment-87746232) (I already updated the jira ticket as well)

~~Also restore all deleted default constructor implementing Serializable (directly or indirectly). https://github.com/Catrobat/Catroid/pull/1205#issuecomment-87750178
LookData, SoundInfo and all XYZBricks implement Serializable and therefore should have a default constructor.~~

I'm not sure about that anymore. First of all it works without default constructors without a problem (https://jenkins.catrob.at/job/Catroid-PullRequest-TestJob/181/)
Second, I finally read http://rolandtapken.de/blog/2010-08/let-xstream-call-default-constructor-where-possible ^^ We don't use this approach. Also it will just handle final transient problems and not even all of them

> This reduces the "final transient" problem to classes without a default constructor.

Third, and the most important one

> XStream is not calling the default constructor during deserialization.

http://xstream.codehaus.org/faq.html#Serialization_no_ctor_running

So, keep the UnnecessaryConstructor and remove all unnecessary constructors?
(I didn't do any more research :wink:)

Pinging all the people! (Do we have an "xstream guy"?)
@avihric @obusher @aried3r @sjaindl @SimonStefan

**Comment by cheidenreich** [comment]
@robertpainsi unfortunately reopening of the old PR was not possible since i force pushed after squashing.

**Comment by robertpainsi** [comment]
@cheidenreich Ah, got it! Didn't know that, thanks for the info!

**Comment by obusher** [comment]
ok to test

**Comment by obusher** [comment]
ok to test

**Comment by obusher** [comment]
ok to test

**Comment by obusher** [comment]
ok to test

**Comment by cheidenreich** [comment]
any update on XStream?

File: catroidTest/src/org/catrobat/catroid/test/utils/Reflection.java [hunk 1]
```diff
@@ -127,8 +127,6 @@ public ParameterList(Object... parameters) {
 				objectClass = int.class;
 			} else if (objectClass == Long.class) {
 				objectClass = long.class;
-			} else if (objectClass == Short.class) {
-				objectClass = short.class;
 			}
```
Comment (obusher) [inline catroidTest/src/org/catrobat/catroid/test/utils/Reflection.java]: Which rule coplained about those two lines? :ear_of_rice:

File: config/pmd_controversial.xml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  -->
+
+<ruleset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         name="Catroid PMD unnecessary ruleset"
+         xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
+         xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 http://pmd.sourceforge.net/ruleset_2_0_0.xsd" >
+	<description >PMD Java Controversial Ruleset for Catroid/PocketCode</description >
+	<rule ref="rulesets/java/controversial.xml" >
+		<exclude name="AssignmentInOperand" />
+        <exclude name="AtLeastOneConstructor" />
```
Comment (obusher) [inline config/pmd_controversial.xml]: Reformat the `<exclude />` tags, most of them are too much on the left side. They should be like the happy one in `line:31`. :walking:

**Comment by obusher** [comment]
Lets open a separate ticket which treats the `~DefaultConstructor` rule, Xstream related research and testing, so we can merge the other rules swifter. :mag:

File: catroidTest/src/org/catrobat/catroid/test/utils/Reflection.java (hunk 1 above)
Comment (robertpainsi) [inline catroidTest/src/org/catrobat/catroid/test/utils/Reflection.java]: http://pmd.sourceforge.net/pmd-4.3.0/rules/controversial.html#AvoidUsingShortType

@cheidenreich
Please undo these changes and exclude the rule. The physics engine (and box2d) is using shorts for collision filtering.

File: catroid/src/org/catrobat/catroid/content/XmlHeader.java [hunk 3]
```diff
@@ -67,94 +67,96 @@
 	@SuppressWarnings("unused")
 	private String userHandle = "";
 
+	@SuppressWarnings("PMD.UnnecessaryConstructor")
 	public XmlHeader() {
+		super();
 	}
 
-	String getProgramName() {
+	public String getProgramName() {
```
Comment (robertpainsi) [inline catroid/src/org/catrobat/catroid/content/XmlHeader.java]: Hmm, I can't find the pmd rule in the docu at the moment, but
- Adding private if the access modifier is missing is a good idea :wink: 
- Adding public if the access modifier is missing isn't a good idea :wink: because there is also package-private.

**Comment by robertpainsi** [comment]
Finally I took a look at the xstream stuff. All approaches have it pros and cons.

Calling the default constructor if there is one, otherwise don't call any constructor, seems a little bit random. This is the approach mentioned in http://rolandtapken.de/blog/2010-08/let-xstream-call-default-constructor-where-possible
I don't like that.

Using the PureJavaReflectionProvider (also mentioned in http://rolandtapken.de/blog/2010-08/let-xstream-call-default-constructor-where-possible), all serializable classes need a default constructor which is called when deserializing an object. This is consistent.

Also the default constructor has to be empty imo. Because when deserializing an object, the default constructor is called first and after that all serialized members are set. So calling a method in the constructor using a serialized member is a paddlin. Setting a serialized member in the constructor is a paddlin.

The last option is the current one. Don't call any constructor when deserializing. This also means that final transient variables won't be initialized, so we can't use the final modifier for transient variables.

In all cases, readResolve is the way to go to do stuff when deserializing an object.

Summarized:
- We can use transient final variables but have to assure that the constructors are and stay empty.

or

- We can not use transient final variables for objects but also don't have to care about constructors.

Any thoughts? opinions? Am I missing something?
ping @avihric

File: catroid/src/org/catrobat/catroid/content/XmlHeader.java (hunk 3 above)
Comment (robertpainsi) [inline catroid/src/org/catrobat/catroid/content/XmlHeader.java]: Found it ^^ It's the DefaultPackage rule. You should exclude it.
http://pmd.sourceforge.net/pmd-5.1.1/rules/java/controversial.html#DefaultPackage

**Comment by robertpainsi** [comment]
> Lets open a separate ticket which treats the ~DefaultConstructor rule, Xstream related research and testing, so we can merge the other rules swifter.

Done, https://jira.catrob.at/browse/CAT-1475

**Comment by obusher** [comment]
@robertpainsi Thank you very much!
@cheidenreich Would be nice if you could polish your PR. ( Rebasing, Squashing to 1 commit....)

**Comment by cheidenreich** [comment]
Too much to rebase and change, i will start over
